source/loader/gamecard.cpp

```diff
@@ -249,58 +1154,147 @@ class Adaptor3DSXToNCCH : public HLE::PXI::FS::File {
             throw std::runtime_error("Can't create/truncate game cards");
         }
 
-        auto ret = file->Open(context, flags);
-        if (ret != std::tie(HLE::OS::RESULT_OK)) {
-            return ret;
+        auto open_ret = file->Open(context, flags);
+        if (open_ret != std::tie(HLE::OS::RESULT_OK)) {
+            return open_ret;
         }
-
-        return ret;
-    }
-
-    FileFormat::Dot3DSX::Header Read3DSXHeader(HLE::PXI::FS::FileContext& context) {
-        // TODO: Use serialization interface instead
-        FileFormat::Dot3DSX::Header dot3dsxheader;
+        
         auto ret = file->Read(context, 0, sizeof(dot3dsxheader), HLE::PXI::FS::FileBufferInHostMemory(dot3dsxheader));
-        if (ret != HLE::OS::OS::ResultAnd<uint32_t> { HLE::OS::RESULT_OK, sizeof(dot3dsxheader) }) {
-            throw std::runtime_error("Failed to read 3DSX header");
+        if (std::get<0>(ret) != HLE::OS::RESULT_OK || std::get<1>(ret) != sizeof(dot3dsxheader)) {
+            throw Mikage::Exceptions::Invalid("Could not read 3DSX header");
         }
+        
+        // Read and validate header(s), populate SMDH data
+        if (dot3dsxheader.header_size == sizeof(FileFormat::Dot3DSX::Header)) {
+            icon_data.magic = { 'S', 'M', 'D', 'H' };
+            std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> conversion;
+            std::u16string longname = conversion.from_bytes(filename ? filename.value() : "Homebrew Application");
+            std::u16string shortname = conversion.from_bytes("Homebrew Application");
+            std::u16string publisher = conversion.from_bytes("Homebrew");
+            
+            for (int i = 0; i < icon_data.application_titles.size(); i++) {
+                ranges::copy(longname, icon_data.application_titles[i].long_description.begin());
+                ranges::copy(shortname, icon_data.application_titles[i].short_description.begin());
```

Comment:
Range APIs support copying to ranges, so I think you don't need to query iterators here:

```suggestion
                ranges::copy(shortname, icon_data.application_titles[i].short_description);
```

(existing code doesn't do this, but let's be better than past-me :))

Reply:
I believe I actually tried this before, but if I remember correctly this doesn't actually work for some reason.